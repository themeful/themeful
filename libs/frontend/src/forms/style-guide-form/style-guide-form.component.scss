:host {
  display: block;
  color: var(--tf-font-color);
}

.form {
  padding: 1rem;
  &__controls {
    display: flex;
    flex-direction: row;
    gap: 8px;
  }
}
