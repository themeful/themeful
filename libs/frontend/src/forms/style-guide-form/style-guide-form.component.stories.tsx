import { h } from '@stencil/core'

export default {
  title: 'Forms/Style Guide',
  args: {
    newMode: {
      formData: {
        fields: {
          baseFontSize: 16,
        },
      },
    },
    editMode: {
      formData: {
        identifier: 'styleGuide1',
        fields: {
          name: 'StyleGuide Name',
          baseFontSize: 16,
        },
      },
    },
  },
}

export const styleGuide = ({ editMode, newMode }) => {
  return (
    <div>
      <div class="header">
        <h1>Style Guide Form</h1>
      </div>
      <div
        class="content"
        style={{
          display: 'grid',
          gridTemplateColumns: '500px 500px',
          gridAutoRows: '300px',
          gap: '1rem',
        }}
      >
        <div class="tf-light" style={{ background: '#fff', padding: '1rem' }}>
          <tf-style-guide-form {...newMode} />
        </div>
        <div style={{ padding: '1rem' }}>
          <tf-style-guide-form {...newMode} />
        </div>
        <div class="tf-light" style={{ background: '#fff', padding: '1rem' }}>
          <tf-style-guide-form {...editMode} />
        </div>
        <div style={{ padding: '1rem' }}>
          <tf-style-guide-form {...editMode} />
        </div>
      </div>
    </div>
  )
}
