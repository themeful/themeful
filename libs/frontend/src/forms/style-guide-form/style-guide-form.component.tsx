import { Component, Event, EventEmitter, h, Prop, State } from '@stencil/core'
import { StyleGuideBase } from '@typings'

@Component({
  tag: 'tf-style-guide-form',
  styleUrl: 'style-guide-form.component.scss',
  shadow: true,
})
export class StyleGuideFormComponent {
  /** Data for the form */
  @Prop() formData: StyleGuideBase

  /** StyleGuide slug */
  @Prop() styleGuide: string

  /** Event emitted when the item is clicked */
  @Event() close: EventEmitter

  @State() changed = false
  @State() name: string
  @State() baseFontSize: number
  @State() editMode: boolean

  componentDidLoad(): void {
    this.editMode = this.styleGuide && true
    this.name = this.formData.name
    this.baseFontSize = this.formData.baseFontSize
  }

  private changeName = (event): void => {
    this.changed = true
    this.name = event.target.value
  }

  private changeFontSize = (event): void => {
    this.changed = true
    this.baseFontSize = event.target.value
  }

  private validate = (): boolean => {
    return this.name.length > 3 && Number(this.baseFontSize) > 0
  }

  private save = (): void => {
    console.log(this.changed, this.validate(), this.name, this.baseFontSize)
    if (this.changed && this.validate()) {
      console.log('save')
    }
  }

  private cancel = (): void => {
    this.close.emit()
  }

  private remove = (): void => {
    if (this.editMode) {
      console.log('remove')
    }
  }

  render() {
    return (
      <div>
        <label>
          Name
          <input value={this.name} onInput={this.changeName} />
        </label>
        <label>
          BaseFontSize
          <input value={this.baseFontSize} onInput={this.changeFontSize} />
        </label>
        <div class="form__controls">
          {this.editMode && (
            <tf-button kind="danger" onClick={this.remove}>
              Delete
            </tf-button>
          )}
          <tf-button kind="secondary" onClick={this.cancel}>
            Cancel
          </tf-button>
          <tf-button kind="primary" onClick={this.save} type="submit">
            Save
          </tf-button>
        </div>
      </div>
    )
  }
}
