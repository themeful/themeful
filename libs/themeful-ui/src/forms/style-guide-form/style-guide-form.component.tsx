import { Component, Event, EventEmitter, h, Prop, State } from '@stencil/core'
import { FormValues, StyleGuideBase, StyleGuideFormAction, StyleGuideFormData } from '@typings'

@Component({
  tag: 'tf-style-guide-form',
  styleUrl: 'style-guide-form.component.scss',
  shadow: true,
})
export class StyleGuideFormComponent {
  /** Data for the form */
  @Prop() formData: StyleGuideFormData

  /** Event emitted when an action is triggered */
  @Event({ composed: false }) action: EventEmitter<StyleGuideFormAction>

  @State() changed = false
  @State() editMode: boolean

  private controls: { [key: string]: HTMLTfTextInputElement } = {}

  public componentWillLoad(): void {
    this.editMode = this.formData.identifier && true
  }

  public componentDidLoad(): void {
    this.action.emit({ action: 'open' })
  }

  private formValues = (): FormValues =>
    Object.entries(this.controls).reduce((result, [key, control]) => {
      result[key] = control.value
      return result
    }, {})

  private validate = (): Promise<boolean> =>
    Promise.all(Object.values(this.controls).map((control) => control.validate())).then(
      (controls) => controls.every((valid) => valid)
    )

  private dirty = (): Promise<boolean> =>
    Promise.all(Object.values(this.controls).map((control) => control.dirty())).then((controls) =>
      controls.some((valid) => valid)
    )

  private save = async (event: Event): Promise<void> => {
    event.preventDefault()
    Promise.all([this.dirty(), this.validate()]).then(([dirty, valid]) => {
      if (dirty && valid) {
        if (this.editMode) {
          this.action.emit({
            action: 'update',
            identifier: this.formData.identifier,
            fields: this.formValues() as unknown as StyleGuideBase,
          })
        } else {
          this.action.emit({
            action: 'create',
            fields: this.formValues() as unknown as StyleGuideBase,
          })
        }
      } else if (valid) {
        this.action.emit({ action: 'close' })
      }
    })
  }

  private cancel = (): void => {
    this.action.emit({ action: 'close' })
  }

  private remove = (): void => {
    if (this.editMode) {
      this.action.emit({ action: 'delete', identifier: this.formData.identifier })
    }
  }

  private baseFontSizeValidation = (value): string | null =>
    Number(value) > 0 ? null : 'Please enter a number'

  public render(): HTMLTfStyleGuideFormElement {
    return (
      <form class="form" onSubmit={this.save}>
        <h3>
          {this.editMode ? 'Edit' : 'Create'}{' '}
          {this.formData.identifier === 'global' ? 'Global' : 'Style Guide'}
        </h3>
        {this.formData.identifier !== 'global' && (
          <tf-text-input
            ref={(el: HTMLTfTextInputElement) => (this.controls['name'] = el)}
            label="Name"
            value={this.formData.fields?.name}
            minLength={4}
          />
        )}
        <tf-text-input
          ref={(el: HTMLTfTextInputElement) => (this.controls['baseFontSize'] = el)}
          label="Base Font Size (px)"
          value={this.formData.fields?.baseFontSize}
          validation={this.baseFontSizeValidation}
        />
        <div class="form__controls">
          {this.editMode && (
            <tf-button kind="danger" onClick={this.remove}>
              Delete
            </tf-button>
          )}
          <tf-button kind="secondary" onClick={this.cancel}>
            Cancel
          </tf-button>
          <tf-button kind="primary" onClick={this.save} type="submit">
            Save
          </tf-button>
        </div>
      </form>
    )
  }
}
