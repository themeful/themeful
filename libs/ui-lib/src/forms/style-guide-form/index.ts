export * from './style-guide-form.component'
