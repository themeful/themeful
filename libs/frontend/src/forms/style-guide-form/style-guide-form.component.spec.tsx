import { newSpecPage } from '@stencil/core/testing'
import { StyleGuideFormComponent } from './style-guide-form.component'

xdescribe('StyleGuideFormComponent', () => {
  it('renders', async () => {
    const { root } = await newSpecPage({
      components: [StyleGuideFormComponent],
      html: '<tf-style-guide-form></tf-style-guide-form>',
    })
    expect(root).toEqualHtml(`
      <tf-style-guide-form>
        <mock:shadow-root>
          <div>
            Hello, World! I'm
          </div>
        </mock:shadow-root>
      </tf-style-guide-form>
    `)
  })

  it('renders with values', async () => {
    const { root } = await newSpecPage({
      components: [StyleGuideFormComponent],
      html: `<tf-style-guide-form first="Stencil" last="'Don't call me a framework' JS"></tf-style-guide-form>`,
    })
    expect(root).toEqualHtml(`
      <tf-style-guide-form first="Stencil" last="'Don't call me a framework' JS">
        <mock:shadow-root>
          <div>
            Hello, World! I'm Stencil 'Don't call me a framework' JS
          </div>
        </mock:shadow-root>
      </tf-style-guide-form>
    `)
  })
})
