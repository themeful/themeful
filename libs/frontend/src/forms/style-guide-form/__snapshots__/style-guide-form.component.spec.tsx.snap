// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleGuideFormComponent renders 1`] = `
<tf-style-guide-form>
  <mock:shadow-root>
    <form class="form">
      <h3>
        Edit Style Guide
      </h3>
      <tf-text-input>
        <mock:shadow-root>
          <label class="text-input">
            <span class="text-input__label">
              Name
            </span>
            <input class="text-input__input" type="text">
            <p class="text-input__hint"></p>
          </label>
        </mock:shadow-root>
      </tf-text-input>
      <tf-text-input>
        <mock:shadow-root>
          <label class="text-input">
            <span class="text-input__label">
              Base Font Size (px)
            </span>
            <input class="text-input__input" type="text">
            <p class="text-input__hint"></p>
          </label>
        </mock:shadow-root>
      </tf-text-input>
      <div class="form__controls">
        <tf-button>
          <mock:shadow-root>
            <button class="button danger" type="button">
              <slot></slot>
            </button>
          </mock:shadow-root>
          Delete
        </tf-button>
        <tf-button>
          <mock:shadow-root>
            <button class="button secondary" type="button">
              <slot></slot>
            </button>
          </mock:shadow-root>
          Cancel
        </tf-button>
        <tf-button>
          <mock:shadow-root>
            <button class="button primary" type="submit">
              <slot></slot>
            </button>
          </mock:shadow-root>
          Save
        </tf-button>
      </div>
    </form>
  </mock:shadow-root>
</tf-style-guide-form>
`;
